sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala

```diff
@@ -403,20 +415,29 @@ class RelationalGroupedDataset protected[sql](
    *
    * {{{
    *   // Compute the sum of earnings for each year by course with each course as a separate column
-   *   df.groupBy($"year").pivot($"course", Seq("dotNET", "Java")).sum($"earnings")
+   *   df.groupBy($"year").pivot($"course", Seq(lit("dotNET"), lit("Java"))).sum($"earnings")
+   * }}}
+   *
+   * For pivoting by multiple columns, use the `struct` function to combine the columns and values:
+   *
+   * {{{
+   *   df
+   *     .groupBy($"year")
+   *     .pivot(struct($"course", $"training"), Seq(struct(lit("java"), lit("Experts"))))
+   *     .agg(sum($"earnings"))
    * }}}
    *
    * @param pivotColumn the column to pivot.
    * @param values List of values that will be translated to columns in the output DataFrame.
    * @since 2.4.0
    */
-  def pivot(pivotColumn: Column, values: Seq[Any]): RelationalGroupedDataset = {
+  def pivot(pivotColumn: Column, values: Seq[Column]): RelationalGroupedDataset = {
```

Comment:
@HyukjinKwon I think this change is better than what https://github.com/apache/spark/pull/21699 did.

Reply:
Hm, wouldn't we better allow this `Seq[Column]` for both `pivot(String ...)` and `pivot(Column ...)` too by `Seq[Any]` since `pivot(String ...)`'s signature allows it?

BTW, we should document this in the param and describe the difference clearly in the documentation. Otherwise, seems the current API change makes the usage potentially quite confusing to me.

Reply:
I am okay if that's what you all guys think. It should really be clearly documented then now if we go ahead with the current way.

Reply:
The very fundamental interface we should have is `pivot(Column, Seq[Column])`, which allows any form and any type of pivot column, and the same with pivot values. This is close to what we support in SQL (SQL pivot support will actually be a subset of DataFrame pivot support after we have this interface), and verifying that the pivot values are constant is taken care of in the Analyzer.
That said, we still need to keep the old `pivot(String, Seq[Any])` for simple usages and for backward compatibility, but I don't think we need to expand its capability. It is pretty clear to me that pivot(String ...) takes a column name and simple objects while with pivot(Column...) you could make any sophisticated use of pivot you would like to.

Reply:
However, we happened to have `pivot(String, Seq[Any])` first which I guess most of users already are used to it.

`pivot(String, Seq[Any])` takes values and columns too (https://github.com/apache/spark/pull/22030#discussion_r208439423, I guess). 
How about we have `pivot(Column, Seq[Any])` takes values and columns too?

Reply:
Here's nothing to argue with the analyzer or something. The first interface exposed is `pivot(String, Seq[Any]`. We better keep it similar to the original version if there isn't a big issue. What's the downside of allowing both by `pivot(Column, Seq[Any])`?

Reply:
The downside is if have both pivot(Column, Seq[Any]) and pivot(Column, Seq[Column]), we would end up having too many versions of pivot and that would be sth confusing I'm afraid.

Reply:
> we would end up having too many versions of pivot and that would be sth confusing I'm afraid.

The previous interface `pivot(Column, Seq[Any])` has existed for more then multiple years. Is this based on actual feedback from users or your speculation?

Reply:
> This assumption of yours is not true. See my reply to your comment below.

By this PR, it's true, right?

Reply:
> The previous interface pivot(Column, Seq[Any]) has existed for more then multiple years. Is this based on actual feedback from users or your speculation?\

This is what @MaxGekk added in https://github.com/apache/spark/pull/21699.

> This assumption of yours is not true. See my reply to your comment below.
No. Seq[Any] takes literal values (objects); Seq[Column] takes `Column` expressions.

Reply:
> This is what @MaxGekk added in #21699.

Oops, I mean `pivot(String, Seq[Any])`. typo

Reply:
> No. Seq[Any] takes literal values (objects); Seq[Column] takes Column expressions.

I mean:
Before:

```
scala> val df = spark.range(10).selectExpr("struct(id) as a")
df: org.apache.spark.sql.DataFrame = [a: struct<id: bigint>]

scala> df.groupBy().pivot("a", Seq(struct(lit(1)))).count().show()
java.lang.RuntimeException: Unsupported literal type class org.apache.spark.sql.Column named_struct(col1, 1)
  at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:78)
  at org.apache.spark.sql.RelationalGroupedDataset$$anonfun$pivot$1.apply(RelationalGroupedDataset.scala:419)
  at org.apache.spark.sql.RelationalGroupedDataset$$anonfun$pivot$1.apply(RelationalGroupedDataset.scala:419)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
  at scala.collection.immutable.List.map(List.scala:296)
  at org.apache.spark.sql.RelationalGroupedDataset.pivot(RelationalGroupedDataset.scala:419)
  at org.apache.spark.sql.RelationalGroupedDataset.pivot(RelationalGroupedDataset.scala:338)
  ... 51 elided

```


After:


```
scala> val df = spark.range(10).selectExpr("struct(id) as a")
df: org.apache.spark.sql.DataFrame = [a: struct<id: bigint>]

scala> df.groupBy().pivot("a", Seq(struct(lit(1)))).count().show()
+---+
|[1]|
+---+
|  1|
+---+
```

Reply:
I think @MaxGekk's intention was to keep the old signature as it is but somehow used "lit" which takes `Column` too. Correct me if I'm wrong, @MaxGekk.
So back to the choice between `pivot(Column, Seq[Column])` and `pivot(Column, Seq[Any])`, I think having an explicit `Seq[Column]` type is less confusing and kind of tells people by itself that we are now support complex types in pivot values.

Reply:
I think https://github.com/apache/spark/pull/22030#discussion_r208456164 makes perfect sense. We really don't need to make it complicated.

>  having an explicit Seq[Column] type is less confusing and kind of tells people by itself that we are now support complex types in pivot values.

My question was that it's from your speculation or actual feedback from users since the original interface has existed for few years and I haven't seen some complaints about this so far as far as I can tell.

It's okay if we clearly document this with some examples. It wouldn't necessarily make some differences between same overloaded APIs.

Reply:
> My question was that it's from your speculation or actual feedback from users...

This is an actual feedback from our users who want to do pivoting by multiple columns. They have to use an external systems (even Microsoft Excel does it better) for pivoting by many columns for now because Spark doesn't allow this. You cannot express for example this on the latest release:
```
trainingSales
      .groupBy($"sales.year")
      .pivot(struct(lower($"sales.course"), $"training"), Seq(
        struct(lit("dotnet"), lit("Experts")),
        struct(lit("java"), lit("Dummies")))
      ).agg(sum($"sales.earnings"))
```

via `def pivot(pivotColumn: String, values: Seq[Any])`. I am not speaking about the recently added method `def pivot(pivotColumn: Column, values: Seq[Any])` which we are going to make more concise and eliminate unnecessary generic type `Any`.

Reply:
I think the point is if users really get confused or not by `Any` because I as a user used this heavily and I have been fine with that for long time. In that case, I thought we better keep it consistent with the original one.